VTCareService/src/main/java/edu/vt/cs/vtcare/vtcareservice/dao/ProviderDao.java

```diff
@@ -38,4 +49,69 @@ public long persistProvider(Provider provider) throws SQLException {
             throw new SQLException("Encountered problem creating a new provider ", e);
         }
     }
+
+    /**
+     *
+     * @param id of provider we're looking for
+     * @return list of providers with only 1 entry, when providerID is known
+     * @throws SQLException on SQL errors like table not existing
+     */
+    public List<Provider> getProvider(long id) throws SQLException {
+        try (PreparedStatement statement = connection.prepareStatement(FIND_BY_PROVIDER_ID_SQL)) {
+            statement.setLong(1, id);
+
+            ResultSet res = statement.executeQuery();
+            List<Provider> items = new ArrayList<>();
+
+            if(res.next() ){
+                extractProvider(res, items);
+            }
+            return items;
+        } catch (SQLException e) {
+            throw new SQLException("Encountered problem fetching a provider ", e);
+        }
+    }
+
+    /**
+     *
+     * @return ListOfAllProviders
+     * @throws SQLException When query is malformed, or other SQL related issues.
+     */
+    public List<Provider> getProviders() throws SQLException{
+        try (PreparedStatement statement = connection.prepareStatement(FIND_ALL_PROVIDERS_SQL)) {
+
+            ResultSet res = statement.executeQuery();
+            List<Provider> items = new ArrayList<>();
+
+            while(res.next() ){
+                extractProvider(res, items);
+            }
+            return items;
+        } catch (SQLException e) {
+            throw new SQLException("Encountered problem fetching providers ", e);
+        }
+    }
+
+    /**
+     *
+     * @param res resultset to iterate through, used to construct provider
+     * @param items data structure used to contain providers
+     * @throws SQLException
+     */
+    private void extractProvider(ResultSet res, List<Provider> items) throws SQLException {
+        Provider prov = new Provider(res.getObject("name", String.class) ,
+                res.getObject("email", String.class) ,
+                "",// probably don't want to leak a provider's password
+                res.getObject("gender", String.class) ,
+                res.getObject("date_of_birth", Date.class).toString(),
+                res.getObject("address", String.class),
+                res.getObject("contact", String.class),
+                res.getObject("experience", Integer.class),
+                res.getObject("specialization", String.class)
+        );
+        prov.setProviderId(res.getObject("id", Integer.class) );
+        items.add(prov);
```

Comment:
It's not wise to pass the list in the parameter, instead we should have the method return the provider object and add the object to the list in the while loop of the calling function

Reply:
I can 100% incorporate this change, but aren't lists pass by reference? Like & in C++?

Reply:
Its just a better practice to return a new object rather than change the referenced objects in parameters.